frontend/catalyst/pennylane_extensions.py

```diff
@@ -935,9 +945,21 @@ def check_is_iterable(x, hint):
         fn = _ensure_differentiable(f)
         grad_params = _check_grad_params(method, scalar_out, h, argnum)
 
-        jaxpr, _ = _make_jaxpr_check_differentiable(fn, grad_params, *params)
+        jaxpr, out_tree = _make_jaxpr_check_differentiable(fn, grad_params, *params)
 
         results = vjp_p.bind(*params, *cotangents, jaxpr=jaxpr, fn=fn, grad_params=grad_params)
+
+        if out_tree.children() != []:
+            func_res = results[: len(jaxpr.out_avals)]
+            vjps = results[len(jaxpr.out_avals) :]
+            func_res = tree_unflatten(out_tree, func_res)
+            # We do not change the shape of the VJP as it is the same as the parameters not
+            # as the output
+            results = tuple([func_res, tuple(vjps)])
+        else:
+            func_res = results[: len(jaxpr.out_avals)]
+            vjps = results[len(jaxpr.out_avals) :]
+            results = tuple([*func_res, tuple(vjps)])
```

Comment:
This structure seems a bit strange no? The function results are expanded but the vjps are in another tuple

Reply:
For the general question: our vjp is very different from Jax vjp https://jax.readthedocs.io/en/latest/_autosummary/jax.vjp.html where they return.

```
res, f_vjp = tuple(res, f_vjp)
```
Here res are unflatten, after that you need to use the function to get the vjps
```
vjps = f_vjp(cot)
```

Reply:
That's true about the vjp, I was mainly thinking of the PyTree behaviour for inputs, outputs, tangents, cotangents, and gradients. Those should ideally all match JAX's version.

Reply:
About the vjp difference, I think we should still return a tuple of `(results, gradients)` just like for the jvp, because like you say we use the same function style for both.